lib/ansible/plugins/connection/ssh.py

```diff
@@ -44,6 +48,119 @@
 SSHPASS_AVAILABLE = None
 
 
+class AnsibleSshConnectionFailure(AnsibleConnectionFailure):
+    ''' the ssh connection plugin had a fatal error '''
+    pass
+
+
+# Could track retry attempts here. A little odd.
+class AnsibleAttemptedSshConnectionFailure(AnsibleConnectionFailure):
+    ''' the ssh connection plugin had a fatal error but may attempt to retry '''
+    pass
+
+class AnsibleSshInitialWriteConnectionFailure(AnsibleSshConnectionFailure):
+    ''' the ssh connection plugin had a fatal error during the initial data write '''
+    pass
+
+
+class AnsibleSshCommandErrorConnectionFailure(AnsibleSshConnectionFailure):
+    ''' the ssh connection plugin had a fatal error (255 return code from ssh command '''
+    def __init__(self, message=None, obj=None, show_content=True, suppress_extended_error=False,
+                 ssh_cmd=None, ssh_stderr=None):
+        message = message or ""
+        super(AnsibleSshCommandErrorConnectionFailure, self).__init__(message=message, obj=obj,
+                                                                      show_content=show_content,
+                                                                      suppress_extended_error=suppress_extended_error)
+        self.ssh_cmd = ssh_cmd
+        self.ssh_stderr = ssh_stderr
+
+        #self.message = '%s\nSTDERR: %s' % (self.message, codecs.escape_decode(self.ssh_stderr))
+        self.message = '%s\n' % (self.message)
+
+class AnsibleSshCommandErrorPermissionDenied(AnsibleSshCommandErrorConnectionFailure):
+    pass
+
+# Example errors
+_example_ssh_errors = """
+@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
+@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+Permissions 0666 for '/home/adrian/.ssh/id_rsa' are too open.
+It is required that your private key files are NOT accessible by others.
+This private key will be ignored.
+Load key "/home/adrian/.ssh/id_rsa": bad permissions
+Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).
+"""
+
+# FIXME: better enum-like
+class SshErrors:
+    permission_denied = 'permission_denied'
+    no_route_to_host = 'no_route_to_host'
+    could_not_resolve = 'could_not_resolve'
+
+class OpenSshStderrErrorMapper:
+    # Attempt to make sense of openssh errors and warnings in hopes of generating
+    # more useful ansible error messages
+
+    # FIXME: The names on the rhs are just placeholders for now, to be replaced with exceptions
+    warning_regexes = {r"""Permissions (.*) for '(.*)' are too open.""":
```

Comment:
The openssh messages searched for here are not translated. (afaict, no openssh messages are translated and no message catalogs are included). 

If a different implementation did translate the message (or just used different messages), the error mapper would fail to find a more specific cause and would fall back to the existing errors. rc=255 will always be a failure, it's just a matter if we can detect a more specific reason. Ditto for the messages that may be libc/os
specific (errno, 'No route to host' etc).

No doubt the regexes will need tuning and updating over time.